### PR TITLE
support ELRSv3 RMFD (maybe)

### PR DESCRIPTION
**Pull request by stronnag**
Support ELRS v3 RMFD (without breaking v2)

### CODE REVIEW

**Comment by stronnag** [comment]
In theory, this supports both V2 and V3 ELRS RMFD. As the developer has neither, testing is appreciated.
Compiled lua [LuaTelemetry_v2.2.0+e3.zip](https://github.com/iNavFlight/OpenTX-Telemetry-Widget/files/10418175/LuaTelemetry_v2.2.0%2Be3.zip)
Source distribution (compiles on radio) [LuaTelemetry_v2.2.0+e3_lua.zip](https://github.com/iNavFlight/OpenTX-Telemetry-Widget/files/10418176/LuaTelemetry_v2.2.0%2Be3_lua.zip)

File: src/SCRIPTS/TELEMETRY/iNav/elrs.lua [hunk 1]
```diff
@@ -3,54 +3,62 @@
...
-		return deviceIsELRS_TX
-	end
+   local offset
+   local id = data[2]
+   local devicesName
+   local majId = -1
+   devicesName, offset = fieldGetString(data, 3)
+   if deviceId == id then
+      if ((fieldGetValue(data,offset,4) == 0x454C5253) and (deviceId == 0xEE)) then
+	 majId = data[offset+8] - 0x30
```
Comment (hexfet) [inline src/SCRIPTS/TELEMETRY/iNav/elrs.lua:40]: This is not the correct data interpretation.  Look how the version is set in Expresslrs CRSF::GetDeviceInformation.

Captures I have prior to elrsv3 don't have the version set (all zeroes), so maybe just check for version 3.  But I can't find the code that sends it in elrs at the moment and out of time to look :)

Here's a device info packet from 3.0.1: `ea 1c 29 ea ee 48 4d 20 45 53 32 34 54 58 00 45 4c 52 53 00 00 00 00 00 03 00 01 1b 00 c9 [Device] 0xee "HM ES24TX^@" 27 parameters`

**Comment by stronnag** [comment]
@hexfet, thanks for reminding me I'd forgotten to push the fixed files.

**Comment by rmaia3d** [comment]
If these latest commits make the files the same as the zips in: https://github.com/iNavFlight/OpenTX-Telemetry-Widget/issues/70#issuecomment-1383191902

Then I can confirm they are working as expected with ELRS v3.1.2 :-)

**Comment by stronnag** [comment]
> If these latest commits make the files the same as the zips in: [#70 (comment)](https://github.com/iNavFlight/OpenTX-Telemetry-Widget/issues/70#issuecomment-1383191902)
> 
> Then I can confirm they are working as expected with ELRS v3.1.2 :-)

They are. I'd forgotten to push them.

**Comment by CapnBry** [comment]
Nice! I see you've got a lot of debug dump in the Lua, and figured it out, but if you need any help with something like this in the future don't hesitate to pop over to the ELRS discord and I'm happy to help with our struct layouts and stuff as well as free code. I [did a similar thing](https://github.com/ExpressLRS/ElrsTelemWidget/blob/main/src/WIDGETS/ELRST/main.lua#L325) a while back for our widget.

* I'd remove the `byteToStr()` function and just use `string.char(b)`. You don't initialize symbolChars so this will never do anything, we just use it to manually convert the up down arrows to the proper character for the proper handset firmware
* The `fieldStrFF()` in `fieldGetString()` is an optimization where we don't rebuild the same string multiple times and run out of memory when refreshing the config list, you can just remove that and the `last` parameter. However, if you're _never_ interested in the device name, you can instead just skip the string builder (looks like you're not using `devicesName` anyway) and just loop until `data[x++] == 0` since they are just null terminated strings. This would eliminate the fieldGetString and byteToString entirely.

**Comment by stronnag** [comment]
Thanks. The code was 'borrowed' from the ELRS Lua; the maintainer does not have CRSF device to do more.

So well informed / tested PRs are *always* welcome.

**Comment by CapnBry** [comment]
> And it's only "Don't Merge" because no one seems interested in testing that it still works on v2. Sad really.

I'm just finishing up for the day but promise you I will test it against v2 and v3 hardware tomorrow. Would you like me to post a diff for removing the unneeded functions and just skipping the deviceName (or "fast-forwarding" as I called it in our code) for inclusion into this as well? If so I can make that change before I test. I'm also a little depressed when I have PRs I spend time on and nobody seems to care about getting them tested and signed off. Honestly I was unaware of this widget until earlier today! I'm happy to help out where I can though.

**Comment by CapnBry** [comment]
Nope this does not work on ExpressLRS 2.x. The code expects that we actually fill the SW Version field but I didn't add that until 3.x. I  was worried stuff that expects it to be a Crossfire version number would break in horrible ways, but then I decided we had to do it because things had to know the version number. I've updated the code so that a 0 == ExpressLRS v2.x. ExpressLRS v1 did not support this config protocol at all so none of this works on it at all. I've also removed any unused functions, updated the link to the annotation from this PR to our protocol docs as that has more complete information.

Tested with iNav 5.1.0 F4, iFlight 2400 RX, RadioMaster TX16S with ExpressLRS versions 2.5.1 and 3.2.0 (but only tested 3 or 4 rates on each version to make sure they were correct).

[elrs_v2v3.patch](https://github.com/iNavFlight/OpenTX-Telemetry-Widget/files/10686240/elrs_v2v3.patch)  I did a patch instead of cloning the whole repo and making my own branch since it is just this one file, hope that's OK but just give me angry eyes if it is not.
```diff
--- src/SCRIPTS/TELEMETRY/iNav/elrs.lua.orig	2023-02-08 08:42:10.182184300 -0500
+++ src/SCRIPTS/TELEMETRY/iNav/elrs.lua	2023-02-08 08:51:22.845087400 -0500
@@ -1,24 +1,10 @@
 -- Adapted from ELRSv2.lua in ExpressLRS
 
-local deviceId = 0xEE
-
-local function byteToStr(b)
-   -- Translate b into a string from symbolChars if available, else use string.char
-   return symbolChars and symbolChars[b] or string.char(b)
-end
-
-local function fieldGetString(data, offset, last)
-   if last then
-      return fieldStrFF(data, offset, last)
-   end
-
-   local result = ""
-   while data[offset] ~= 0 do
-      result = result .. byteToStr(data[offset])
-      offset = offset + 1
-   end
-
-   return result, offset + 1
+local function fieldStrFF(data, offset)
+  while data[offset] ~= 0 do
+    offset = offset + 1
+  end
+  return offset + 1
 end
 
 local function fieldGetValue(data, offset, size)
@@ -29,49 +15,19 @@
    return result
 end
 
---[[
-local function debug_devinfo(data)
-   local ut = getRtcTime()
-   local s = string.format("/LOGS/E_%d.txt", ut)
-   local fh = io.open(s,"wb")
-   dbg_serialize(fh, data)
-   io.close(fh)
-end
-
-local function dbg_serialize (fh, o)
-   if type(o) == "number" then
-      io.write(fh, o)
-   elseif type(o) == "string" then
-      io.write(fh, string.format("%q", o))
-   elseif type(o) == "table" then
-      io.write(fh, "{\n")
-      for k,v in pairs(o) do
-	 io.write(fh, "  [")
-	 dbg_serialize(fh, k)
-	 io.write(fh, "] = ")
-	 dbg_serialize(fh, v)
-	 io.write(fh, ",\n")
-      end
-      io.write(fh, "}\n")
-   else
-      print("cannot serialize a " .. type(o))
-   end
-end
---]]
-
---  See https://github.com/iNavFlight/OpenTX-Telemetry-Widget/issues/70#issuecomment-1383190840
+--  See https://github.com/ExpressLRS/ExpressLRS/wiki/CRSF-Protocol#device-info--device-ping-response-0x29
 --  For annotation
-
 local function parseDeviceInfoMessage(data)
-   local offset
-   local id = data[2]
-   local devicesName
    local majId = -1
-   devicesName, offset = fieldGetString(data, 3)
-   if deviceId == id then
+   local deviceId = data[2]
+   local offset = fieldStrFF(data, 3)
+   if deviceId == 0xEE then -- TX device
       -- debug_devinfo(data)
-      if ((fieldGetValue(data,offset,4) == 0x454C5253) and (deviceId == 0xEE)) then
-	 majId = data[offset+9]
+      if fieldGetValue(data,offset,4) == 0x454C5253 then -- 'ELRS'
+        majId = data[offset+9]
+        if majId == 0 then
+          majId = 2 -- v2 did not report version, v1 didn't support this protocol
+        end
       end
    end
    return majId
```

**Comment by stronnag** [comment]
Thanks very much. I'll apply and merge. Much appreciated.

**Pull request merged**